setup/src/Magento/Setup/Module/Di/Code/Reader/ClassesScanner.php

```diff
@@ -43,7 +80,16 @@ public function addExcludePatterns(array $excludePatterns)
      */
     public function getList($path)
     {
+
         $realPath = realpath($path);
+        $isGeneration = $this->isGeneration($realPath);
+
+        // Generation folders should not have their results cached since they may actually change during compile
+        if (!$isGeneration) {
+            if (isset($this->fileResults[$realPath])) {
```

Comment:
Instead of 2 nested ifs I would be better to  have them as !$isGenerated && isset